### PR TITLE
LibreSSL 3.7 adds support for Ed25519

### PR DESCRIPTION
**Pull request by botovq**
This brings support for the raw public key API.

### CODE REVIEW

**Comment by botovq** [comment]
I have tested this on an OpenBSD dev system by enabling `ed25519_supported()` and `x25519_supported()` in the backend of cryptography 3.8.3. For Ed25519 all tests pass, for X25519 seven tests fail since they assume presence of the `EVP_PKEY_get1_tls_encodedpoint()` API. There are currently no plans to support that in LibreSSL.

I did not include the test enabling for LibreSSL in this PR since it is not entirely clear to me how you'd like to handle that.

**Comment by reaperhulk** [comment]
Have you had a chance to test this against the 3.7.0 development work? I see we can't add this to CI right now since 3.7 isn't available (although perhaps we should consider adding portable tip-of-tree to our CI like we've done for boring and openssl...)

**Comment by botovq** [comment]
Yes, see above.

**Comment by reaperhulk** [comment]
Oops, you replied right before I sent my question, sorry! If it passes ed25519 I'm okay with enabling it in the tests -- that way we'll automatically start testing against it as soon as we add 3.7 to CI. Looks like we need to revisit the APIs we use in x25519 to see whether we can switch there as well.

**Comment by botovq** [comment]
Would it be acceptable to redefine:

```
#define CRYPTOGRAPHY_OPENSSL_LESS_THAN_111B \
    (OPENSSL_VERSION_NUMBER < 0x10101020 || CRYPTOGRAPHY_LIBRESSL_LESS_THAN_370)
```

This would enable Ed25519 and Ed448 support and the latter could be disabled by excluding LibreSSL as already done for BoringSSL.

We could release LibreSSL 3.7.0 fairly soon for dev work and testing (in a few weeks).

**Comment by reaperhulk** [comment]
I'm reluctant to make `CRYPTOGRAPHY_OPENSSL_LESS_THAN_111B` have special logic compared to other defines with similar names, but we actually only use it to gate this single feature. I think we can use your proposal, but rename the value to `CRYPTOGRAPHY_HAS_WORKING_ED25519` or similar with a comment reminding us that it's tied to OpenSSL 1.1.1b and LibreSSL 3.7 (and in the future when those are old and we drop support this define will no longer be needed).

**Comment by botovq** [comment]
Thanks, that makes sense. I'll do that later today or tomorrow since I'm travelling. I'll relearn out how to run the tests from your HEAD and will update this PR with something along these lines. I'll also let you know when the changes are synced to our GH repo.

**Comment by botovq** [comment]
I was going to iterate on it but then life happened, as usual...
Let me try to finish it up in the coming days.

**Comment by alex** [comment]
No worries, just wanted to make sure we were on top of it.

On Tue, Nov 22, 2022 at 7:54 AM Theo Buehler ***@***.***> wrote:
>
> I was going to iterate on it but then life happened, as usual...
> Let me try to finish it up in the coming days.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.Message ID: ***@***.***>



-- 
All that is necessary for evil to succeed is for good people to do nothing.

File: src/_cffi_src/openssl/cryptography.py [hunk 1]
```diff
@@ -68,6 +71,10 @@
 #else
 #define CRYPTOGRAPHY_NEEDS_OSRANDOM_ENGINE 0
 #endif
+/* Ed25519 support is available from OpenSSL 1.1.1b and LibreSSL 3.7.0. */
+#define CRYPTOGRAPHY_HAS_WORKING_ED25519 \
+    (OPENSSL_VERSION_NUMBER >= 0x10101020 && \
```
[the remark this answers is not in the document]
Comment (botovq) [reply src/_cffi_src/openssl/cryptography.py]: Done.

File: src/cryptography/hazmat/backends/openssl/backend.py [hunk 2]
```diff
@@ -2043,8 +2043,9 @@ def ed448_supported(self) -> bool:
         if self._fips_enabled:
             return False
         return (
-            not self._lib.CRYPTOGRAPHY_OPENSSL_LESS_THAN_111B
+            self._lib.CRYPTOGRAPHY_HAS_WORKING_ED25519
```
[the remark this answers is not in the document]
Comment (botovq) [reply src/cryptography/hazmat/backends/openssl/backend.py]: I removed that part of the commit. I thought the idea was to garbage collect LESS_THAN_111B, but I see your point.

File: src/_cffi_src/openssl/cryptography.py [hunk 3]
```diff
@@ -68,6 +71,10 @@
 #else
 #define CRYPTOGRAPHY_NEEDS_OSRANDOM_ENGINE 0
 #endif
+/* Ed25519 support is available from OpenSSL 1.1.1b and LibreSSL 3.7.0. */
+#define CRYPTOGRAPHY_HAS_WORKING_ED25519 \
+    (!CRYPTOGRAPHY_OPENSSL_LESS_THAN_111B || \
+    !CRYPTOGRAPHY_LIBRESSL_LESS_THAN_370)
```
Comment (alex) [inline src/_cffi_src/openssl/cryptography.py:77]:
```suggestion
    (CRYPTOGRAPHY_IS_LIBRESSL && !CRYPTOGRAPHY_LIBRESSL_LESS_THAN_370))
```

Comment (botovq) [reply src/_cffi_src/openssl/cryptography.py:77]: Yes, just noticed by the test failure. Will fix.

**Pull request merged**